### PR TITLE
[Merged by Bors] - feat(RepresentationTheory/GroupCohomology): Noether's generalization of Hilbert's Theorem 90

### PR DESCRIPTION
**Pull request by 101damnations**
---

This PR shows that given a finite extension of fields $L/K,$ $H^1(Aut_K(L), L^\times)$ is trivial. Not sure about naming & whether this theorem should have its own file or not.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+/-- Given `f : Aut_K(L) → Lˣ`, the sum `∑ f(φ) • φ` for `φ ∈ Aut_K(L)`, as a function `L → L`. -/
+noncomputable def aux (f : (L ≃ₐ[K] L) → Lˣ) : L → L :=
+  Finsupp.total (L ≃ₐ[K] L) (L → L) L (fun φ => φ)
+    (Finsupp.equivFunOnFinite.symm (fun φ => (f φ : L)))
+
+theorem aux_ne_zero (f : (L ≃ₐ[K] L) → Lˣ) : aux K L f ≠ 0 :=
+/- the set `Aut_K(L)` is linearly independent in the `L`-vector space `L → L`, by Dedekind's
+linear independence of characters -/
+  have : @LinearIndependent (L ≃ₐ[K] L) L (L → L) (fun f => f) _ _ _ :=
```
Comment (riccardobrasca) [inline Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean:62]: Can you use `(... := ...)` instead of `@` here? It is usually more readable.

**Review by riccardobrasca — APPROVED** [review]

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  have h := linearIndependent_iff.1 this
+    (Finsupp.equivFunOnFinite.symm (fun φ => (f φ : L)))
+  fun H => Units.ne_zero (f 1) (FunLike.ext_iff.1 (h H) 1)
+
+end Hilbert90
+section
+open Hilbert90
+variable {K L : Type*} [Field K] [Field L] [Algebra K L] [FiniteDimensional K L]
+
+theorem hilbert90 (f : (L ≃ₐ[K] L) → Lˣ)
```
[the remark this answers is not in the document]
Comment (joelriou) [reply Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean:75]: In order to use the lemma, I think it is more convenient if we can specify the cocycle `f` explicitly, as it is done a few lines below. I would suggest keeping `(...)` here.

File: Mathlib/Algebra/Group/Equiv/TypeTags.lean [hunk 3]
```diff
@@ -107,6 +107,20 @@ abbrev MulEquiv.toAdditive'' [AddZeroClass G] [MulOneClass H] :
...
+/-- Multiplicative equivalence between multiplicative endomorphisms of a `MulOneClass` `M`
+and additive endomorphisms of `Additive M`. -/
+@[simps] def monoidEndToAdditive (M : Type*) [MulOneClass M] :
+    Monoid.End M ≃* AddMonoid.End (Additive M) :=
+  { MonoidHom.toAdditive with
+    map_mul' := fun _ _ => rfl }
+
+/-- Multiplicative equivalence between additive endomorphisms of an `AddZeroClass` `A`
+and multiplicative endomorphisms of `Multiplicative A`. -/
+@[simps] def addMonoidEndToMultiplicative (A : Type*) [AddZeroClass A] :
```
[the remark this answers is not in the document]
Comment (101damnations) [reply Mathlib/Algebra/Group/Equiv/TypeTags.lean:119]: I don't think so; I get an error I'm not sure how to fix, and none of the other similar pairs of definitions in this file use `to_additive` either.

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+## TODO
+
+* The original Hilbert's Theorem 90, deduced from the cohomology of general finite cyclic groups.
+* Develop Galois cohomology to extend Noether's result to infinite Galois extensions.
+* "Additive Hilbert 90": let `L/K` be a finite Galois extension. Then $H^n(Gal(L/K), L)$ is trivial
+for all $1 ≤ n.$
+
+-/
+suppress_compilation
```
[the remark this answers is not in the document]
Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean:47]: Not really; only the final instance needs to be `noncomputable` if I remove it.

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean (hunk 1 above)
Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean]: thanks, done

File: Mathlib/Algebra/Algebra/Equiv.lean [hunk 5]
```diff
@@ -754,6 +754,13 @@ instance applyMulSemiringAction : MulSemiringAction (A₁ ≃ₐ[R] A₁) A₁ w
   mul_smul _ _ _ := rfl
 #align alg_equiv.apply_mul_semiring_action AlgEquiv.applyMulSemiringAction
 
+instance : MulDistribMulAction (A₁ ≃ₐ[R] A₁) A₁ˣ where
```
[the remark this answers is not in the document]
Comment (101damnations) [reply Mathlib/Algebra/Algebra/Equiv.lean]: Thanks, generalised and moved it

[the remark this answers is not in the document]
Comment (101damnations) [reply Mathlib/Algebra/Algebra/Equiv.lean:757]: huh, I thought I tested that the action of Aut gets inferred from the action of End earlier but apparently not, guess I'll add it back

File: Mathlib/RepresentationTheory/Basic.lean [hunk 6]
```diff
@@ -290,7 +291,34 @@ theorem ofMulAction_single (g : G) (x : H) (r : k) :
...
 end MulAction
+section DistribMulAction
 
+variable (M A : Type*) [Monoid M] [AddCommGroup A] [DistribMulAction M A]
+
+/-- Turns an `AddCommGroup` `A` with a `DistribMulAction` of a monoid `M` into a
+`ℤ`-linear `M`-representation on `A`. -/
+def ofDistribMulAction : Representation ℤ M A :=
+  (addMonoidEndRingEquivInt A : _ →* _).comp (DistribMulAction.toAddMonoidEnd M A)
+
```
Comment (kbuzzard) [inline Mathlib/RepresentationTheory/Basic.lean:302]: Would it also be useful to have a def of `Representation R M A` given some commutative ring R acting compatibly on M and A? Then maybe typeclass inference can fill in the actions if R=Z?

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+import Mathlib.FieldTheory.Galois
+import Mathlib.RepresentationTheory.GroupCohomology.LowDegree
+import Mathlib.LinearAlgebra.LinearIndependent
+
+/-!
+# Hilbert's Theorem 90
+
+Let `L/K` be a finite extension of fields. Then this file proves Noether's generalization of
+Hilbert's Theorem 90: that the 1st group cohomology $H^1(Aut_K(L), Lˣ)$ is trivial. We state it
+both in terms of $H^1$ and otherwise.
```
Comment (kbuzzard) [inline Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean:15]:
```suggestion
both in terms of $H^1$ and in terms of cocycles being coboundaries.
```

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+Hilbert's original statement was that if $L/K$ is Galois, and $Gal(L/K)$ is cyclic, generated
+by an element `σ`, then for every `x : L` such that $N_{L/K}(x) = 1,$ there exists `y : L` such
+that $x = y/σ(y).$ This can be deduced from the fact that the function $Gal(L/K) → L^\times$
+sending $σ^i \mapsto xσ(x)σ^2(x)...σ^{i-1}(x)$ is a 1-cocycle. Alternatively, we can derive it by
+analyzing the cohomology of finite cyclic groups in general.
+
+Noether's generalization also holds for infinite Galois extensions.
+
+## Main definitions
```
Comment (kbuzzard) [inline Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean:25]:
```suggestion
## Main statements
```

**Review by kbuzzard — COMMENTED** [review]
Very nice!

File: Mathlib/RepresentationTheory/Basic.lean (hunk 6 above)
Comment (101damnations) [reply Mathlib/RepresentationTheory/Basic.lean]: Thanks, that works

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean (hunk 7 above)
Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean]: Thanks, done

File: Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean (hunk 8 above)
Comment (101damnations) [reply Mathlib/RepresentationTheory/GroupCohomology/Hilbert90.lean]: Thanks, done

**Comment by riccardobrasca** [comment]
Thanks!

bors d+

**Pull request closed without merging**